File: src/core/peer_dbs.py
"""
@package core
peer_dbs module
"""

# -*- coding: iso-8859-15 -*-

# This code is distributed under the GNU General Public License (see
# THE_GENERAL_GNU_PUBLIC_LICENSE.txt for extending this information).
# Copyright (C) 2014, the P2PSP team.
# http://www.p2psp.org

# DBS: Data Broadcasting Set of rules

# {{{

from __future__ import print_function
import threading
import sys
import socket
import struct
import time

from core.common import Common

from core.color import Color
from core._print_ import _print_
#from core.peer_ims import Peer_IMS
from core.peer_ims_gui import Peer_IMS_GUI as Peer_IMS

# }}}

# Some useful definitions.
ADDR = 0
PORT = 1

def _p_(*args, **kwargs):
    """Colorize the output."""
    if __debug__:
        sys.stdout.write(Common.DBS_COLOR)
        _print_("DBS:", *args)
        sys.stdout.write(Color.none)

class Peer_DBS(Peer_IMS):
    # {{{

    # {{{ Class "constants"

    MAX_CHUNK_DEBT = 128 # Peer's rejecting threshold

    LOGGING = False # A IMS???
    LOG_FILE = ""   # A IMS???

    # }}}

    def __init__(self, peer):
        # {{{

        _p_("max_chunk_debt =", self.MAX_CHUNK_DEBT)
        _p_("Initialized")

        # }}}

    def say_hello(self, node):
        # {{{

        self.team_socket.sendto(b'H', node)
        _p_("[Hello] sent to %s" % str(node))

        # }}}

    def say_goodbye(self, node):
        # {{{

        self.team_socket.sendto(b'G', node)
        _p_("[Goodbye] sent to %s" % str(node))

        # }}}

    def receive_magic_flags(self):
        self.magic_flags = struct.unpack("B",self.splitter_socket.recv(struct.calcsize("B")))[0]
        _p_("Magic flags =", bin(self.magic_flags))

    def receive_the_number_of_peers(self):
        # {{{

        self.debt = {}      # Chunks debts per peer.
        self.peer_list = [] # The list of peers structure.

        #sys.stdout.write(Color.green)
        _p_("Requesting the number of monitors and peers to", self.splitter_socket.getpeername())
        self.number_of_monitors = socket.ntohs(struct.unpack("H",self.splitter_socket.recv(struct.calcsize("H")))[0])
        _p_("The number of monitors is", self.number_of_monitors)
        self.number_of_peers = socket.ntohs(struct.unpack("H",self.splitter_socket.recv(struct.calcsize("H")))[0])
        _p_("The size of the team is", self.number_of_peers, "(apart from me)")

        #sys.stdout.write(Color.none)

        # }}}

    def receive_the_list_of_peers(self):
        # {{{

        self.debt = {}      # Chunks debts per peer.
        self.peer_list = [] # The list of peers structure.

        #sys.stdout.write(Color.green)
        _p_("Requesting", self.number_of_peers, "peers to", self.splitter_socket.getpeername())
        #number_of_peers = socket.ntohs(struct.unpack("H",self.splitter_socket.recv(struct.calcsize("H")))[0])
        #_print_("The size of the team is", number_of_peers, "(apart from me)")

        tmp = self.number_of_peers
        while tmp > 0:
            message = self.splitter_socket.recv(struct.calcsize("4sH"))
            IP_addr, port = struct.unpack("4sH", message) # Ojo, !H ????
            IP_addr = socket.inet_ntoa(IP_addr)
            port = socket.ntohs(port)
            peer = (IP_addr, port)
            _p_("[hello] sent to", peer)
            self.say_hello(peer)
            if __debug__:
                _p_("[%5d]" % tmp, peer)
            else:
                _print_("{:.2%}\r".format((self.number_of_peers-tmp)/self.number_of_peers), end='')

            self.peer_list.append(peer)
            self.debt[peer] = 0
            tmp -= 1

        _p_("List of peers received")
        #sys.stdout.write(Color.none)

        # }}}

    def receive_my_endpoint(self):
        # {{{

        message = self.splitter_socket.recv(struct.calcsize("4sH"))
        IP_addr, port = struct.unpack("4sH", message) # Ojo, !H ????
        IP_addr = socket.inet_ntoa(IP_addr)
        port = socket.ntohs(port)
        self.me = (IP_addr, port)
        _p_("me =", self.me)

        # }}}

    def listen_to_the_team(self):
        # {{{ Create "team_socket" (UDP) as a copy of "splitter_socket" (TCP)

        self.create_team_socket()
        try:
            # In Windows systems this call doesn't work!
            self.team_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        except Exception as e:
            print ("DBS:", e)
            pass
        self.team_socket.bind(('', self.splitter_socket.getsockname()[PORT]))

        # This is the maximum time the peer will wait for a chunk
        # (from the splitter or from another peer).
        self.team_socket.settimeout(1)

        # }}}

    def process_message(self, message, sender):
        # {{{ Now, receive and send.

        if len(message) == struct.calcsize(self.message_format):
            # {{{ A video chunk has been received

            chunk_number, chunk = self.unpack_message(message)
            self.chunks[chunk_number % self.buffer_size] = chunk
            self.received_flag[chunk_number % self.buffer_size] = True
            self.received_counter += 1

            if sender == self.splitter:
                # {{{ Send the previous chunk in burst sending
                # mode if the chunk has not been sent to all
                # the peers of the list of peers.

                # {{{ debug

                _p_(self.team_socket.getsockname(), "<-", chunk_number, "-", sender)
                if __debug__: # No aqui. Tal vez, DIS

                    if self.LOGGING:
                        self.log_message("buffer correctnes {0}".format(self.calc_buffer_correctnes()))
                        self.log_message("buffer filling {0}".format(self.calc_buffer_filling()))

                # }}}

                while( (self.receive_and_feed_counter < len(self.peer_list)) and (self.receive_and_feed_counter > 0) ):
                    peer = self.peer_list[self.receive_and_feed_counter]
                    self.team_socket.sendto(self.receive_and_feed_previous, peer)
                    self.sendto_counter += 1

                    _p_(self.team_socket.getsockname(), "-",\
                        socket.ntohs(struct.unpack(self.message_format, \
                                                   self.receive_and_feed_previous)[0]),\
                        "->", peer)

                    self.debt[peer] += 1
                    if self.debt[peer] > self.MAX_CHUNK_DEBT:
                        _p_(peer, 'removed by unsupportive (' + str(self.debt[peer]) + ' lossess)')
                        del self.debt[peer]
                        self.peer_list.remove(peer)

                    self.receive_and_feed_counter += 1

                self.receive_and_feed_counter = 0
                self.receive_and_feed_previous = message

               # }}}
            else:
                # {{{ The sender is a peer

                _p_(self.team_socket.getsockname(), \
                    "<-", chunk_number, "-", sender)

                if sender not in self.peer_list:
                    # The peer is new
                    self.peer_list.append(sender)
                    self.debt[sender] = 0
                    _p_(sender, 'added by chunk', chunk_number)
                else:
                    self.debt[sender] -= 1

                # }}}

            # {{{ A new chunk has arrived and the
            # previous must be forwarded to next peer of the
            # list of peers.
            if ( self.receive_and_feed_counter < len(self.peer_list) and ( self.receive_and_feed_previous != '') ):
                # {{{ Send the previous chunk in congestion avoiding mode.

                peer = self.peer_list[self.receive_and_feed_counter]
                self.team_socket.sendto(self.receive_and_feed_previous, peer)
                self.sendto_counter += 1

                self.debt[peer] += 1
                if self.debt[peer] > self.MAX_CHUNK_DEBT:
                    _p_(peer, 'removed by unsupportive (' + str(self.debt[peer]) + ' lossess)')
                    del self.debt[peer]
                    self.peer_list.remove(peer)

                _p_(self.team_socket.getsockname(), "-", \
                    socket.ntohs(struct.unpack(self.message_format, self.receive_and_feed_previous)[0]),\
                    "->", peer)

                self.receive_and_feed_counter += 1

                # }}}
            # }}}

            return chunk_number

            # }}}
        else:
            # {{{ A control chunk has been received
            _p_("Control message received")
            if message == 'H':
                if sender not in self.peer_list:
                    # The peer is new
                    self.peer_list.append(sender)
                    self.debt[sender] = 0
                    _p_(sender, 'added by [hello]')
            else:
                if sender in self.peer_list:
                    #sys.stdout.write(Color.red)
                    _p_(self.team_socket.getsockname(), '\b: received "goodbye" from', sender)
                    #sys.stdout.write(Color.none)
                    self.peer_list.remove(sender)
                    del self.debt[sender]
            return -1

            # }}}

        # }}}

    def keep_the_buffer_full(self):
        # {{{

        Peer_IMS.keep_the_buffer_full(self)
        if (self.played_chunk % self.debt_memory) == 0:
            for i in self.debt:
                self.debt[i] /= 2

        #_p_("Number of peers in the team:", len(self.peer_list)+1)
        #_p_(self.team_socket.getsockname(),)
        #if __debug__:
        #    for p in self.peer_list:
        #        print ("DBS:", p,)
        #    print ()

        # }}}

    def polite_farewell(self):
        # {{{

        _p_('Goodbye!')
        for x in range(3):
            self.process_next_message()
            self.say_goodbye(self.splitter)
        for peer in self.peer_list:
            self.say_goodbye(peer)

        # }}}

    def buffer_data(self):
        # {{{

        # Number of times that the previous received chunk has been sent
        # to the team. If this counter is smaller than the number
        # of peers in the team, the previous chunk must be sent in the
        # burst mode because a new chunk from the splitter has arrived
        # and the previous received chunk has not been sent to all the
        # peers of the team. This can happen when one or more chunks
        # that were routed towards this peer have been lost.
        self.receive_and_feed_counter = 0

        # This "private and static" variable holds the previous chunk
        # received from the splitter. It is used to send the previous
        # received chunk in the congestion avoiding mode. In that
        # mode, the peer sends a chunk only when it received a chunk
        # from another peer or from the splitter.
        self.receive_and_feed_previous = ""

        self.sendto_counter = 0

        self.debt_memory = 1 << self.MAX_CHUNK_DEBT

        Peer_IMS.buffer_data(self)

        # }}}

    def run(self):
        # {{{

        #Peer_IMS.peers_life(self)
        Peer_IMS.run(self)
        self.polite_farewell()

        # }}}

    def am_i_a_monitor(self):
        # {{{
        if self.number_of_peers < self.number_of_monitors:
            # Only the first peers of the team are monitor peers
            return True
        else:
            return False
        #message = self.splitter_socket.recv(struct.calcsize("c"))
        #if struct.unpack("c", message)[0] == '1':
        #    return True
        #else:
        #    return False

        # }}}

    # }}}

    # The following methods should be inheritaged ... in DIS??

    def calc_buffer_correctnes(self):
        zerochunk = struct.pack("1024s", "0")
        goodchunks = badchunks = 0
        for i in range(self.buffer_size):
            if self.received_flag[i]:
                if self.chunks[i] == zerochunk:
                    badchunks += 1
                else:
                    goodchunks += 1
        return goodchunks / float(goodchunks + badchunks)

    def calc_buffer_filling(self):
        chunks = 0
        for i in range(self.buffer_size):
            if self.received_flag[i]:
                chunks += 1
        return chunks / float(self.buffer_size)

    def log_message(self, message):
        self.LOG_FILE.write(self.build_log_message(message) + "\n")
        #print >>self.LOG_FILE, self.build_log_message(message)

    def build_log_message(self, message):
        return "{0}\t{1}".format(repr(time.time()), message)

File: src/core/peer_dbs.cc
//
//  peer_dbs.h - Data Broadcasting Set of rules.
//
//  This code is distributed under the GNU General Public License (see
//  THE_GENERAL_GNU_PUBLIC_LICENSE.txt for extending this information).
//  Copyright (C) 2016, the P2PSP team.
//  http://www.p2psp.org
//

#include "peer_dbs.h"

namespace p2psp {

  Peer_DBS::Peer_DBS() {
    // {{{

    max_chunk_debt_ = kMaxChunkDebt;
    ready_to_leave_the_team_ = false;
      
    // }}}
  }

  Peer_DBS::~Peer_DBS() {}

  void Peer_DBS::Init() {
    // {{{

#if defined __DEBUG__ || defined __SORS__
    TRACE("Initialized");
#endif
#if defined __DEBUG__ || defined __PARAMS__
    //TRACE("max_chunk_debt = " + std::to_string(kMaxChunkDebt));
    TRACE("max_chunk_debt = "
	  << max_chunk_debt_);
#endif

    // }}}
  }

  std::vector<ip::udp::endpoint> *Peer_DBS::GetPeerList() {
    return &peer_list_;
  }

  /*void Peer_DBS::SetTeamPort(uint16_t team_port) {
    team_port_ = team_port;
  }

  uint16_t Peer_DBS::GetTeamPort() {
    return team_port_;
  }

  uint16_t Peer_DBS::GetDefaultTeamPort() {
    return kTeamPort;
    }*/

  // Unused in this layer!
  void Peer_DBS::ReceiveMyEndpoint() {
    boost::array<char, 6> buffer;
    char *raw_data = buffer.data();
    ip::address ip_addr;
    ip::udp::endpoint peer;
    int port;

    read(splitter_socket_, ::buffer(buffer));
    in_addr ip_raw = *(in_addr *)(raw_data);
    ip_addr = ip::address::from_string(inet_ntoa(ip_raw));
    port = ntohs(*(short *)(raw_data + 4));

    me_ = ip::udp::endpoint(ip_addr, port);

#if defined __DEBUG__ || defined __TRAFFIC__
    TRACE("me = ("
	  << me_.address().to_string()
	  << ","
          << std::to_string(me_.port())
	  << ")");
#endif
    
  }

  void Peer_DBS::SayHello(const ip::udp::endpoint &node) {
    // {{{

    std::string hello("H");

    team_socket_.send_to(buffer(hello), node);

#if defined __DEBUG__ || defined __CHURN__
    TRACE("[Hello] sent to "
          << "(" << node.address().to_string()
	  << ","
          << std::to_string(node.port())
	  << ")");
#endif
    
    // }}}
  }

  void Peer_DBS::SayGoodbye(const ip::udp::endpoint &node) {
    // {{{

    std::string goodbye("G");

    team_socket_.send_to(buffer(goodbye), node);

#if defined __DEBUT__ || defined __TRAFFIC__
    TRACE("[Goodbye] sent to "
          << "(" << node.address().to_string()
	  << ","
          << std::to_string(node.port())
	  << ")");
#endif
    
    // }}}
  }

  /*void Peer_DBS::ReceiveMagicFlags() {
    std::vector<char> magic_flags(1);
    read(splitter_socket_, ::buffer(magic_flags));
    TRACE("Magic flags = " << std::bitset<8>(magic_flags[0]));

    if (this->magic_flags_ != magic_flags[0]) {
      ERROR("The splitter has different magic flags ("
        << std::bitset<8>(magic_flags[0]) << ") than this peer ("
        << std::bitset<8>(magic_flags_) << ").");
      ERROR("Please run splitter with a different parameter, or compile peer "
        << "with a different set of rules.");
      exit(1);
    }
  }*/

  void Peer_DBS::ReceiveTheNumberOfPeers() {
    // {{{

    boost::array<char, 2> buffer;

#if defined __DEBUG__ || defined __TRAFFIC__
    // sys.stdout.write(Color.green)
    TRACE("Requesting the number of monitors and peers to ("
          << splitter_socket_.remote_endpoint().address().to_string()
	  << ","
          << std::to_string(splitter_socket_.remote_endpoint().port())
	  << ")");
#endif
    
    read(splitter_socket_, ::buffer(buffer));
    number_of_monitors_ = ntohs(*(short *)(buffer.c_array()));
#if defined __DEBUG__ || defined __TRAFFIC__
    TRACE("The number of monitors = "
	  << number_of_monitors_);
#endif
    
    read(splitter_socket_, ::buffer(buffer));
    number_of_peers_ = ntohs(*(short *)(buffer.c_array()));
#if defined __DEBUG__ || defined __TRAFFIC__
    TRACE("Total number of peers in the team (excluding me) = "
	  << number_of_peers_);
#endif
    
    // }}}
  }

  void Peer_DBS::ReceiveTheListOfPeers() {
    // {{{

    boost::array<char, 6> buffer;
    char *raw_data = buffer.data();
    ip::address ip_addr;
    ip::udp::endpoint peer;
    int port;

    ReceiveTheNumberOfPeers();
    
#if defined __DEBUG__ || defined __TRAFFIC__
    // sys.stdout.write(Color.green)
    TRACE("Requesting"
	  << number_of_peers_
	  << " peers to ("
          << splitter_socket_.remote_endpoint().address().to_string()
          << ","
	  << std::to_string(splitter_socket_.remote_endpoint().port())
          << ")");
#endif
    
    // number_of_peers =
    // socket.ntohs(struct.unpack("H",self.splitter_socket.recv(struct.calcsize("H")))[0])
    //_print_("The size of the team is", number_of_peers, "(apart from me)")

    int tmp = number_of_peers_;
    while (tmp > 0) {
      read(splitter_socket_, ::buffer(buffer));
      in_addr ip_raw = *(in_addr *)(raw_data);
      ip_addr = ip::address::from_string(inet_ntoa(ip_raw));
      port = ntohs(*(short *)(raw_data + 4));

      peer = ip::udp::endpoint(ip_addr, port);
#if defined __DEBUG__ ||  defined __CHURN__
      TRACE("[hello] sent to ("
	    << peer.address().to_string()
	    << ","
            << std::to_string(peer.port())
	    << ")");
#endif
      SayHello(peer);

      /*#if defined __DEBUG_TRAFFIC__
      TRACE(std::to_string((number_of_peers_ - tmp) / number_of_peers_));
      #endif*/
      
      peer_list_.push_back(peer);
      debt_[peer] = 0;
      tmp--;
      //}
    }

#if defined __DEBUG__ || defined __TRAFFIC__
    TRACE("List of peers received");
#endif

    // }}}
  }

  void Peer_DBS::ListenToTheTeam() {
    // {{{

#if defined __DEBUG__ || defined __PARAMS__
    TRACE("splitter port = "
	  << splitter_socket_.local_endpoint().port());
#endif
    
    ip::udp::endpoint endpoint(ip::address_v4::any(),
                               splitter_socket_.local_endpoint().port());

    team_socket_.open(endpoint.protocol());
    team_socket_.set_option(ip::udp::socket::reuse_address(true));
    team_socket_.bind(endpoint);

    // This is the maximum time the peer will wait for a chunk
    // (from the splitter or from another peer).
    team_socket_.set_option(socket_base::linger(true, 30));

    // }}}

  }

  int Peer_DBS::ProcessMessage(const std::vector<char> &message,
			       const ip::udp::endpoint &sender) {
    // {{{
    
    // Now, receive and send.

      /*#if defined __DEBUG_TRAFFIC__
    TRACE("Size: "
	  << message.size()
	  << " vs "
	  << message_size_)
	  #endif*/
    
      // TODO: remove hardcoded values
      if (message.size() == message_size_) {
	// {{{

	// A video chunk has been received
	ip::udp::endpoint peer;
	
	uint16_t chunk_number = ntohs(*(uint16_t *) message.data());

	chunk_ptr[chunk_number % buffer_size_].data
	  = std::vector<char>(message.data() + sizeof(uint16_t), message.data() + message.size());
	chunk_ptr[chunk_number % buffer_size_].chunk_number = chunk_number /*% buffer_size_*/;

	/*chunks_[chunk_number % buffer_size_] = {
	  std::vector<char>(message.data() + sizeof(uint16_t),
			    message.data() + sizeof(uint16_t) + chunk_size_),
			    true};*/
	
#if defined __DEBUG__ || defined __BUFFERING__
	TRACE("Chunk Inserted at: "
	      << (chunk_number % buffer_size_));
#endif
	
	received_counter_++;

	if (sender == splitter_) {
	  // {{{

	  // Send the previous chunk in burst sending
	  // mode if the chunk has not been sent to all
	  // the peers of the list of peers.

#if defined __DEBUG__ || defined __TRAFFIC__
	  TRACE("("
		<< team_socket_.local_endpoint().address().to_string()
		<< ","
		<< std::to_string(team_socket_.local_endpoint().port())
		<< ")"
		<< "<-"
		<< std::to_string(chunk_number)
		<< "-"
		<< "("
		<< sender.address().to_string()
		<< ","
		<< std::to_string(sender.port())
		<< ")");
#endif
	  
	  while (receive_and_feed_counter_ < (int) peer_list_.size()
		 && (receive_and_feed_counter_ > 0 or modified_list_)) {
	    // {{{

	    peer = peer_list_[receive_and_feed_counter_];
	    
	    team_socket_.send_to(buffer(receive_and_feed_previous_), peer);
	    sendto_counter_++;

#if defined __DEBUG__ || defined __TRAFFIC__
	    TRACE("("
		  << team_socket_.local_endpoint().address().to_string()
		  << ","
		  << std::to_string(team_socket_.local_endpoint().port())
		  << ")"
		  << "-"
		  << std::to_string(ntohs(*(uint16_t *)receive_and_feed_previous_.data()))
		  << "->"
		  << "(" << peer.address().to_string()
		  << ","
		  << std::to_string(peer.port())
		  << ")");
#endif

	    debt_[peer]++;
	      
	    if (debt_[peer] > max_chunk_debt_/*kMaxChunkDebt*/) {
#if defined __DEBUG__ || defined __CHURN__
	      TRACE("("
		    << peer.address().to_string()
		    << ","
		    << std::to_string(peer.port())
		    << ")"
		    << " removed by unsupportive (" + std::to_string(debt_[peer]) + " lossess)");
#endif
	      debt_.erase(peer);
	      peer_list_.erase(std::find(peer_list_.begin(), peer_list_.end(), peer));
	      //receive_and_feed_counter_--;
	    }
	    
	    receive_and_feed_counter_++;

	    // }}}
	  }

	  // Start playback with the first chunk received from the splitter //
	  std::vector<char> empty(message_size_, 0);
	  if (receive_and_feed_previous_ == empty) {
	    // {{{

	    played_chunk_ = ntohs(*(uint16_t *)message.data());
#if defined __DEBUG__ || defined __BUFFERING__
	    TRACE("First chunk to play modified "
		  << std::to_string(played_chunk_));
#endif

	    // }}}
	  }
	  // --------------------------------------------- //
	    
	  modified_list_ = false;

#if defined __DEBUG__ || defined __TRAFFIC__
	  TRACE("Sent "
		<< receive_and_feed_counter_
		<< " of "
		<< peer_list_.size());
	  TRACE("Last Chunk saved in receive and feed: "
		<< ntohs(*(uint16_t *)message.data()));
#endif
	  receive_and_feed_counter_ = 0;
	  receive_and_feed_previous_ = message;

	  // }}}
	} else {
	  // {{{

#if defined __DEBUG__ || defined __TRAFFIC__
	  TRACE("("
		<< team_socket_.local_endpoint().address().to_string()
		<< ","
		<< std::to_string(team_socket_.local_endpoint().port())
		<< ")"
		<< "<-"
		<< std::to_string(chunk_number)
		<< "-"
		<< "(" << sender.address().to_string()
		<< ","
		<< std::to_string(sender.port())
		<< ")");
#endif
	  if (peer_list_.end() == std::find(peer_list_.begin(), peer_list_.end(), sender)) {
	    // {{{

	    peer_list_.push_back(sender);
	    debt_[sender] = 0;
#if defined __DEBUG__ || defined __CHURN__
	    TRACE("("
		  << sender.address().to_string()
		  << ","
		  << std::to_string(sender.port())
		  << ")"
		  << " added by chunk "
		  << std::to_string(chunk_number));
#endif

	    // }}}
	  } else {
	    // {{{

	    debt_[sender]--;

	    // }}}
	  }

	  // }}}
	}
	
	// A new chunk has arrived and the previous must be forwarded
	// to next peer of the list of peers.
	
	std::vector<char> empty(message_size_, 0);
	
	if (receive_and_feed_counter_ < (int) peer_list_.size() && receive_and_feed_previous_!=empty) {
	  // {{{

	  // Send the previous chunk in congestion avoiding mode.
	  
	  peer = peer_list_[receive_and_feed_counter_];
#if defined __DEBUG__ || defined __TRAFFIC__
	  TRACE("Sending in congestion avoiding mode to "
		<< peer.address().to_string()
		<< ":"
		<< peer.port()
		<< "/ counter "
		<< receive_and_feed_counter_);
#endif
	  team_socket_.send_to(buffer(receive_and_feed_previous_), peer);
	  sendto_counter_++;
	  
	  debt_[peer]++;
	  
	  if (debt_[peer] > max_chunk_debt_/*kMaxChunkDebt*/) {
	    // {{{

#if defined __DEBUG__ || defined __CHURN__
	    TRACE("("
		  << peer.address().to_string()
		  << ","
		  << std::to_string(peer.port())
		  << ")"
		  << " removed by unsupportive (" + std::to_string(debt_[peer]) + " lossess)");
#endif
	    debt_.erase(peer);
	    peer_list_.erase(std::find(peer_list_.begin(), peer_list_.end(), peer));
	    //receive_and_feed_counter_--;

	    // }}}
	  }

#if defined __DEBUG__ || defined __TRAFFIC__
	  TRACE("("
		<< team_socket_.local_endpoint().address().to_string()
		<< ","
		<< std::to_string(team_socket_.local_endpoint().port())
		<< ")"
		<< "-"
		<< std::to_string(ntohs(*(short *)receive_and_feed_previous_.data()))
		<< "->"
		<< "("
		<< peer.address().to_string()
		<< ","
		<< std::to_string(peer.port())
		<< ")");
#endif
	  
	  receive_and_feed_counter_++;

	  // }}}
	}
	
	return chunk_number;

	// }}}
      } else {
	// {{{

	// A control chunk has been received
#if defined __DEBUG__ || defined __CHURN__
	TRACE("Control message received");
#endif
	if (message[0] == 'H') {
	  // {{{

	  if (peer_list_.end() == std::find(peer_list_.begin(), peer_list_.end(), sender)) {
	    // {{{

	    // The peer is new
	    peer_list_.push_back(sender);
	    debt_[sender] = 0;
#if defined __DEBUG__ || defined __CHURN__
	    TRACE("("
		  << sender.address().to_string()
		  << ","
		  << std::to_string(sender.port())
		  << ")"
		  << " added by [hello] ");
#endif

	    // }}}
	  }

	  // }}}
	} else {
	  // {{{

	  if (peer_list_.end() != std::find(peer_list_.begin(), peer_list_.end(), sender)) {
	    // {{{

#if defined __DEBUG__ || defined __CHURN__
	    TRACE("("
		  << team_socket_.local_endpoint().address().to_string()
		  << ","
		  << std::to_string(team_socket_.local_endpoint().port())
		  << ") \b: received \"goodbye\" from ("
		  << sender.address().to_string()
		  << ","
		  << std::to_string(sender.port())
		  << ")");
#endif
	    // sys.stdout.write(Color.none)
	    peer_list_.erase(std::find(peer_list_.begin(), peer_list_.end(),sender));
	    debt_.erase(sender);
	    if (receive_and_feed_counter_ > 0){
	      // {{{

	      modified_list_ = true;
	      receive_and_feed_counter_--;

	      // }}}
	    }

	    // }}}
	  } else {
	    // {{{

	    if (sender == splitter_){
	      // {{{

#if defined __DEBUG__ || defined __CHURN__
	      TRACE("Goodbye received from splitter");
#endif
	      waiting_for_goodbye_ = false;

	      // }}}
	    }

	    // }}}
	  }

	  // }}}
	}
	
	return -1;

	// }}}
      }
    
    return -1;

    // }}}
  }

  float Peer_DBS::CalcBufferCorrectness() {
    // {{{

    std::vector<char> zerochunk(chunk_size_, 0);

    int goodchunks = 0;
    int badchunks = 0;

    for (std::vector<Chunk>::iterator it = chunks_.begin(); it != chunks_.end();
         ++it) {
      if (it->chunk_number != -1) {
        if (it->data == zerochunk) {
          badchunks++;
        } else {
          goodchunks++;
        }
      }
    }

    return goodchunks / (float)(goodchunks + badchunks);

    // }}}
  }

  float Peer_DBS::CalcBufferFilling() {
    // {{{

    int chunks = 0;

    for (std::vector<Chunk>::iterator it = chunks_.begin(); it != chunks_.end();
         ++it) {
      if (it->chunk_number != -1) {
        chunks++;
      }
    }

#if defined __DEBUG__ || defined __BUFFERING__
    TRACE("Chunks in Buffer: " << chunks)
#endif
    return chunks / (float)buffer_size_;

    // }}}
  }

  void Peer_DBS::PoliteFarewell() {
    // {{{

    std::vector<char> message(message_size_);
    ip::udp::endpoint sender;

#if defined __DEBUG__ || defined __CHURN__
    TRACE("Goodbye!");
#endif
    
    while (receive_and_feed_counter_ < (int) peer_list_.size()) {
      team_socket_.send_to(buffer(receive_and_feed_previous_), peer_list_[receive_and_feed_counter_]);
      team_socket_.receive_from(buffer(message), sender);
#if defined __DEBUG__ || defined __TRAFFIC__
      TRACE("("
	    << team_socket_.local_endpoint().address().to_string()
	    << ","
	    << std::to_string(team_socket_.local_endpoint().port())
	    << ")"
	    << "-"
	    << std::to_string(ntohs(*(uint16_t*)receive_and_feed_previous_.data()))
	    << "->"
	    << "("
	    << peer_list_[receive_and_feed_counter_].address().to_string()
	    << ","
	    << std::to_string(peer_list_[receive_and_feed_counter_].port()) << ")");
#endif
      receive_and_feed_counter_++;
    }
    
    for (std::vector<ip::udp::endpoint>::iterator it = peer_list_.begin(); it != peer_list_.end(); ++it) {
      SayGoodbye(*it);
      //team_socket_.receive_from(buffer(message), sender);
    }

    ready_to_leave_the_team_ = true;
#if defined __DEBUG__ || defined __CHURN__
    TRACE("Ready to leave the team!");
#endif
    
    // }}}
  }

  void Peer_DBS::BufferData() {
    // {{{

    // Number of times that the previous received chunk has been sent to the team.
    // If this counter is smaller than the number of peers in the team, the
    // previous chunk must be sent in the burst mode because a new chunk from the
    // splitter has arrived and the previous received chunk has not been sent to
    // all the peers of the team. This can happen when one or more chunks that
    // were routed towards this peer have been lost.
    receive_and_feed_counter_ = 0;

    // This "private and static" variable holds the previous chunk received from
    // the splitter. It is used to send the previous received chunk in the
    // congestion avoiding mode. In that mode, the peer sends a chunk only when it
    // received a chunk from another peer or from the splitter.
    receive_and_feed_previous_ = std::vector<char>(message_size_);

    sendto_counter_ = 0;

    debt_memory_ = 1 << max_chunk_debt_/*kMaxChunkDebt*/;

    waiting_for_goodbye_ = true;

    Peer_core::BufferData();

    // }}}
  }

  void Peer_DBS::Start() {
    // {{{

    thread_group_.interrupt_all();
    thread_group_.add_thread(new boost::thread(&Peer_DBS::Run, this));

    // }}}
  }

  void Peer_DBS::Run() {
    // {{{

    std::vector<char> message(message_size_);
    ip::udp::endpoint sender;
    
    while (player_alive_ or waiting_for_goodbye_) {
      KeepTheBufferFull();
      
      if (!player_alive_){
	//for (int i = 0; i < 1; i++) {
	// ProcessNextMessage();
	SayGoodbye(splitter_);
	//team_socket_.receive_from(buffer(message), sender);
	//}
      }
      //PlayNextChunk();
    }
    PoliteFarewell();

    // }}}
  }

  bool Peer_DBS::AmIAMonitor() {
    // {{{

    // Only the first peers of the team are monitor peers
    return number_of_peers_ < number_of_monitors_;

    // message = self.splitter_socket.recv(struct.calcsize("c"))
    // if struct.unpack("c", message)[0] == '1':
    //    return True
    // else:
    //    return False

    // }}}
  }

  int Peer_DBS::GetNumberOfPeers() {
    // {{{

    return number_of_peers_;

    // }}}
  }

  void Peer_DBS::SetMaxChunkDebt(int max_chunk_debt) {
    // {{{

    max_chunk_debt_ = max_chunk_debt;

    // }}}
  }

  int Peer_DBS::GetMaxChunkDebt() {
    // {{{

    return max_chunk_debt_;

    // }}}
  }

  int Peer_DBS::GetDefaultMaxChunkDebt() {
    // {{{

    return kMaxChunkDebt;

    // }}}
  }

  bool Peer_DBS::IsReadyToLeaveTheTeam(){
    // {{{
    
    return ready_to_leave_the_team_;

    // }}}
  }
}
